### ISSUE TITLE
Support Of Traversable By List Type

### ISSUE THREAD

**Issue by XedinUnknown** [vimeo/psalm#5022]
Psalm [declares support for `list` type][1], which stands for a zero-based series without gaps. Logically, a `Traversable` that obeys the same rules is also a list, not just array. [Weirdly][2], Psalm doesn't complain when phpDoc overrides `iterable` typehint to `list<int>`; and yet it does complain when giving an `Iterator` (`ArrayIterator`) with a list inside.

Is this supported? Or maybe semi-supported? If not, how about adding support for that? The consumer of the function doesn't care if it's an array or a `Traversable`, as long as it's a sequential zero-based list with no gaps.

[1]: https://psalm.dev/docs/annotating_code/type_syntax/array_types/#lists
[2]: https://psalm.dev/r/9aaaf5c0c6

**Comment by orklah** [comment]
This would need to be a separate type than list because https://3v4l.org/0CCIX

Even an ArrayIterator can't be used as an array in all situations

**Comment by orklah** [comment]
Also, your example is not weird. list is a subtype of array itself a subtype of iterable. 

Psalm allow being more precise in phpdoc than in signature but will crash when trying to push a Traversable when a list is expected because list imply array (and ArrayIterator is not an array)

**Comment by XedinUnknown** [comment]
Thanks for the quick response!

Yes, of course a `list` cannot be treated as an `array`, because a `list` IMHO defines how it is _consumed_, not how it is _configured_. In other words, this pseudo-type would declare an immutable interface and impose some conditions on the keys, and that's it. If it is not defined that way, maybe it should be re-defined. Because if I do `$list['asd'] = 123`, then it is no longer a list.

**Comment by orklah** [comment]
The meaning of list can't be changed now. That would imply breaking thousand of builds for very little added value.

**Comment by XedinUnknown** [comment]
Well, then its a first ever type that changes when its members change :)

On a serious note, what would you do?

**Comment by orklah** [comment]
Well, you haven't produced any sample of code so I'm not sure what your issue really is, but if I assume that you have an ArrayIterator that you want to pass as a list, I'd do one of the following:
- use https://www.php.net/iterator_to_array and then document the resulting array as a list
- use other assertions in your code to make psalm understand your offset exists

If a Psalm type is really needed, it'd be something like `list-iterable` that would be a subtype of `iterable` and a supertype of `array`. However, by the little I can grasp of Psalm, it'd be non-trivial and a consequent amount of work to get rid of all the edge cases.
It'd also probably imply the creation of a `non-empty-list-iterable` that would have all sorts of weird interactions with other related types

**Comment by orklah** [comment]
> Yes, of course a `list` cannot be treated as an `array`, because a `list` IMHO defines how it is _consumed_, not how it is _configured_. In other words, this pseudo-type would declare an immutable interface and impose some conditions on the keys, and that's it. If it is not defined that way, maybe it should be re-defined. Because if I do `$list['asd'] = 123`, then it is no longer a list.

Re-reading your post, I may have misunderstood your point. `$list['asd'] = 123` can't happen on a `list` because psalm will either detect it's illegal or immediately change the type to a standard array. A list is just a special case of an array that happen to be really useful for all sorts of assumptions(we can iterate on it with `for` more easily for example). But it's still an array for all intents and purposes and it will be accepted in any place an array is expected. An ArrayIterator is not an array and (unfortunately) can't be pushed to array_sum or similar functions. 
If the meaning of list were to be changed to include ArrayIterator, array_sum would need to stop accepting lists or Psalm would need to accept that it can't detect when we pass a Traversable to array_sum anymore. 
This is the breakage I was talking about. If list can't be accepted by array_* functions anymore, simple code like `array_sum(explode('|', '1|2));` is broken in Psalm

**Comment by XedinUnknown** [comment]
Yep, I get your point about arrays and breaking. But my point is that conceptually, a list is any sequential zero-based series of things. An iterator can be sequential and zero-based just the same. When we talk about a list in general, we only imply that, IMHO. We don't imply that it is countable - the list can be infinite. We don't imply that we can access random elements. We just imply that we can list the items of the list, one after another, in the order of their appearance. For a consumer of a list, the only thing that they need to do is that. Any additional properties of the list, like countability, are properties of the sub-types of the list. If they are added to the super-type, that breaks ISP, and the consumer of the list is imposing useless requirements on its consumer. `array_sum()` wont work with lists, because its internals require an array, as is suggested by its name. An algorithm that sums the elements of a list, as I imagine it to be, would not require anything else of it, and will simply itearate over the items and add them all together. In that case, maybe we will see an `list_sum()` function, or a change in the implementation of `array_sum()`, or a userland function that does this. But it's besides the point, because the concept of a list has IMHO a very narrow definition, and `Traversable` fits into it just as well as `array`.

Due to the lack of generics and intersection types in PHP, I've had to create an [interop standard][`dhii/collections-interface`] - specifically to have ISP-compliant lists. In my applications, when I need have something that can be interpreted as a list I use `Traversable`, and of course that's where the problem lies: `Traversable` does not satisfy `list`, and so I have to necessarily choose between `Traversable`, which means that I require consumers to wrap their simple arrays into more complex array iterators, and a subtype of `array`  - the `list`, and then I cannot have Generators or other kinds of objects. This is very inconvenient. I can also use the `iterable` hint to allow both objects and arrays, but this does not impose the rules of `list` onto them. If I needed a finite list that can be counted, I would depend on a [`CountableListInterface`][] instead. If I also wanted random access, well then that list would also be a map, and so I would depend on [`CountableMapInterface`][].

Does this make sense to you?

[`CountableListInterface`]: https://github.com/Dhii/collections-interface/blob/develop/src/CountableListInterface.php
[`CountableMapInterface`]: https://github.com/Dhii/collections-interface/blob/develop/src/CountableMapInterface.php
[`dhii/collections-interface`]: https://github.com/Dhii/collections-interface

**Comment by orklah** [comment]
I see. However, the fact that list does nearly what you want doesn't mean we can just change list to conform with that need. The Psalm definition of list is really a "continuous, integer-indexed arrays like ["red", "yellow", "blue"] .

These arrays have the property $arr === array_values($arr), and represent a large percentage of all array usage in PHP applications."

The very definition of a list needs array_values() to be described.

You could maybe do something with a plugin that could add a `list-iterable` that you could ship with your library but I'm not sure how powerful Plugins really are and how far you'll be able to go.

Maybe the easier thing would be to have a toList() method in your collections? Whenever the list concept is not needed, you can use the collection as an object, but whenever you need to fallback to an array-list, you call toList()?

Do you have a concrete code example where this is an issue for the user of your library?

**Comment by weirdan** [comment]
Psalm's `list` is just a different data structure and is not what you (and CS in general) call lists. Much like PHP's `array` is not a true array in common CS sense. As unfortunate as that may be the ship has sailed, the name is taken, so if you want to fit your lists into Psalm type system you would have to choose another name.

With that said, I don't think being 0-based is a useful constraint for iterators because it would be immediately violated on the first iteration step. E.g.:
```php
function makeListlikeIterable(): iterable {
    yield from [1, 2, 3];
}

function acceptsListlike($param): void {}

$it = makeListlikeIterable();
// $it is supposedly listlike because it would yield these keys: 0, 1, 2
foreach ($it as $elt) {
    acceptsListlike($it); // but here it is not anymore.
}
```
If `acceptsListlike()` required `$param->key()` to be 0, which looks sensible given your definition, Psalm would have to consider `$it` not a listlike as soon as iteration starts. Is such easily spoiled type actually useful in practice?

On the other hand, a relaxed version of the constraint that just guaranties keys to be sequential does indeed look useful.

**Comment by weirdan** [comment]
> Well, then its a first ever type that changes when its members change :)

The type itself does not change, it's the variable that changes, acquiring a different type. Totally legal in PHP as it's a dynamically typed language. Much like it's totally legal to do `$i = PHP_INT_MAX; $i += 1;` and have `$i`s type to become `float`.

**Comment by XedinUnknown** [comment]
All fair points, thanks! So, would you consider adding a type like `seq` that would just mean a sequencial series `iterator`?

**Comment by XedinUnknown** [comment]
In fact, for iteration of a list, the key is irrelevant. If the consumers wants a key-value pair, then that's what each element should expose. If they want to track the current iteration number, then they should add a counter. So, I support that something that simply promises a sequencial `iterator` is useful and well-defined :+1:.

**Comment by muglug** [comment]
> All fair points, thanks! So, would you consider adding a type like seq that would just mean a sequencial series iterator?

No, I don't think that's particularly valuable for a wide number of users

**Comment by XedinUnknown** [comment]
Oh. Sad. Any advice on how I could add this in userland code?

**Comment by orklah** [comment]
Maybe if you had a concrete example of code with the issue on psalm.dev

**Comment by XedinUnknown** [comment]
@orklah, what is missing from my [original example][1]?

[1]: https://psalm.dev/r/9aaaf5c0c6

**Comment by orklah** [comment]
Well if you remove the wrong `list<int>` in phpdoc there is no issue anymore so I'm not sure what it's supposed to show.

**Comment by weirdan** [comment]
'list' in Psalm's interpretation has been officially blessed for PHP 8.1: https://wiki.php.net/rfc/is_list

**Comment by orklah** [comment]
Well, not completely. @TysonAndre idea at first was a `is_list` function: https://externals.io/message/112560

It was later renamed as `array_is_list`. It would stay coherent if tomorrow, PHP implemented an `iterable_is_list` or even a `is_list` with a different definition (or even a fully fleshed out list type)

It is awesome news for static analysis though as list as a type in documentation should become widely used

**Comment by XedinUnknown** [comment]
@orklah, removing it would be wrong, because I expect a list of integers, and provide a list of integers. You can probably remove most type errors by removing the type hint, but is this what we should be doing?

I'm not saying it's an issue of Psalm. I'm saying that Psalm does not support a very basic and intiutive aspect of a list, neither with `list` nor with any other type, because Psalm's `list` adds more expectations onto the type than necessary.

So, I am asking how I could add support for a perhaps new type in Psalm, without modifying Psalm itself.

**Comment by orklah** [comment]
Ok, maybe this was not clear enough.

Wherever you put a list expectation in your code and push an iterable, this will be an issue because that's simply not how Psalm works and you won't be able to keep `list` for this usage, even with a plugin.

Now, this means that you'll have to remove `list` in your code. To help you, I would need to have an example of why you want a new type - that won't be list - and I need to see a code that would break unless you use this new type.

Because in your previous example, the `iterable` seems more than enough
